diff --git a/Makefile b/Makefile
--- luvit.orig/Makefile	2011-10-26 18:03:32.000000000 +0400
+++ luvit/Makefile	2011-10-27 20:28:14.976629625 +0400
@@ -58,8 +58,9 @@
 
 ${LUADIR}/src/libluajit.a:
 	git submodule update --init ${LUADIR}
-	sed -e "s/#XCFLAGS+= -DLUAJIT_ENABLE_LUA52COMPAT/XCFLAGS+= -DLUAJIT_ENABLE_LUA52COMPAT/" -i deps/luajit/src/Makefile
-	sed -e "s/#XCFLAGS+= -DLUA_USE_APICHECK/XCFLAGS+= -DLUA_USE_APICHECK/" -i deps/luajit/src/Makefile
+	# N.B. Mac has incompatible sed -i syntax
+	sed -e "s/#XCFLAGS+= -DLUAJIT_ENABLE_LUA52COMPAT/XCFLAGS+= -DLUAJIT_ENABLE_LUA52COMPAT/" deps/luajit/src/Makefile >deps/luajit/src/Makefile.new && mv -f deps/luajit/src/Makefile.new deps/luajit/src/Makefile
+	sed -e "s/#XCFLAGS+= -DLUA_USE_APICHECK/XCFLAGS+= -DLUA_USE_APICHECK/" deps/luajit/src/Makefile >deps/luajit/src/Makefile.new && mv -f deps/luajit/src/Makefile.new deps/luajit/src/Makefile
 	$(MAKE) -C ${LUADIR}
 
 ${UVDIR}/uv.a:
diff --git a/lib/stack.lua b/lib/stack.lua
index eee24b3..b73ceeb 100644
--- a/lib/stack.lua
+++ b/lib/stack.lua
@@ -59,11 +59,12 @@ function Stack.mount(mountpoint, ...)
 
   local stack = Stack.compose(...)
 
-  if mountpoint:sub(#mountpoint) == "/" then
-    mountpoint = mountpoint:sub(1, #mountpoint - 1)
-  end
+  --if mountpoint:sub(#mountpoint) == "/" then
+  --  mountpoint = mountpoint:sub(1, #mountpoint - 1)
+  --end
 
-  local matchpoint = mountpoint .. "/"
+  --local matchpoint = mountpoint .. "/"
+  local matchpoint = mountpoint
 
   return function(req, res, continue)
     local url = req.url
@@ -91,6 +92,7 @@ local Debug = require('debug')
 function Stack.errorHandler(req, res, err)
   if err then
     res:set_code(500)
+    debug(Debug.traceback(err))
     res:finish(Debug.traceback(err) .. "\n")
     return
   end
