diff --git a/lib/stack.lua b/lib/stack.lua
index eee24b3..b73ceeb 100644
--- a/lib/stack.lua
+++ b/lib/stack.lua
@@ -59,11 +59,12 @@ function Stack.mount(mountpoint, ...)
 
   local stack = Stack.compose(...)
 
-  if mountpoint:sub(#mountpoint) == "/" then
-    mountpoint = mountpoint:sub(1, #mountpoint - 1)
-  end
+  --if mountpoint:sub(#mountpoint) == "/" then
+  --  mountpoint = mountpoint:sub(1, #mountpoint - 1)
+  --end
 
-  local matchpoint = mountpoint .. "/"
+  --local matchpoint = mountpoint .. "/"
+  local matchpoint = mountpoint
 
   return function(req, res, continue)
     local url = req.url
@@ -91,6 +92,7 @@ local Debug = require('debug')
 function Stack.errorHandler(req, res, err)
   if err then
     res:set_code(500)
+    debug(Debug.traceback(err))
     res:finish(Debug.traceback(err) .. "\n")
     return
   end
