--- tests.js.orig	2011-10-26 19:41:34.000000000 +0400
+++ tests.js	2011-10-27 20:22:28.462911372 +0400
@@ -5,7 +5,7 @@
 protocols = ['xhr-streaming', 'iframe-eventsource', 'iframe-htmlfile', 'xhr-polling', 'iframe-xhr-polling', 'jsonp-polling'];
 //protocols = ['websocket'];
 protocols = ['websocket', 'xhr-streaming', 'iframe-eventsource', 'iframe-htmlfile', 'xhr-polling', 'iframe-xhr-polling', 'jsonp-polling'];
-protocols = ['iframe-eventsource'];
+//protocols = ['iframe-eventsource'];
 newSockJS = function(path, protocol) {
   var url;
   url = /^http/.test(path) ? path : client_opts.url + path;
@@ -302,7 +302,7 @@
   test_protocol_messages(protocol);
 }
 module('other');
-/*test("amending url", function() {
+test("amending url", function() {
   var dl, r, t;
   dl = document.location;
   r = new SockJS('//blah:1/abc', []);
@@ -370,7 +370,7 @@
     a = new Array(counter);
     go = function() {
       return SockJS.chunkingTest(client_opts.url + '/echo', function(r) {
-        if ($.browser.msie && $.browser.version < 8) {
+        if ($.browser.msie && $.browser.version < 8 && !window.performance) {
           equal(r, false);
         } else {
           equal(r, true);
@@ -431,4 +431,4 @@
 for (_j = 0, _len2 = protocols.length; _j < _len2; _j++) {
   protocol = protocols[_j];
   test_protocol_errors(protocol);
-}*/
+}
